File: 11_binary_search/01_nth_root.py
def NthRoot(n: int, m: int) -> int:
    low, high = 0, m

    while low <= high:
        mid = (low + high) // 2

        curr = mid**n

        if curr == m:
            return mid
        elif curr < m:
            low = mid + 1
        else:
            high = mid - 1

    return -1

File: 11_binary_search/01_nth_root.rs
fn nth_root(n: i32, m: i32) -> i32 {
    let (mut low, mut high) = (0, m);

    while low <= high {
        let mid = (low + high) / 2;
        let curr = mid.pow(n as u32);

        if curr == m {
            return mid;
        } else if curr < m {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    -1
}

fn main() {
    assert_eq!(nth_root(3, 27), 3);
    assert_eq!(nth_root(4, 69), -1);
}
